barretenberg/cpp/src/barretenberg/stdlib/primitives/group/cycle_scalar.cpp

```diff
@@ -82,126 +100,122 @@ template <typename Builder> cycle_scalar<Builder> cycle_scalar<Builder>::create_
-        // By this point, we know that limb1 fits in the range `1 << BigScalarField::NUM_LIMB_BITS to  (1 <<
-        // BigScalarField::NUM_LIMB_BITS) + limb1_max.get_maximum_value() we need to slice this limb into 2. The first
-        // is LO_BITS - BigScalarField::NUM_LIMB_BITS (which reprsents its contribution to *this.lo) and the second
-        // represents the limbs contribution to *this.hi Step 1: compute the max bit sizes of both slices
-        const size_t lo_bits_in_limb_1 = LO_BITS - BigScalarField::NUM_LIMB_BITS;
-        const size_t hi_bits_in_limb_1 = (static_cast<size_t>(limb1_max.get_msb()) + 1) - lo_bits_in_limb_1;
-
-        // Step 2: compute the witness values of both slices
-        const uint256_t limb_1 = limb1.get_value();
-        const uint256_t limb_1_hi_multiplicand = (uint256_t(1) << lo_bits_in_limb_1);
-        const uint256_t limb_1_hi_v = limb_1 >> lo_bits_in_limb_1;
-        const uint256_t limb_1_lo_v = limb_1 - (limb_1_hi_v << lo_bits_in_limb_1);
-
-        // Step 3: instantiate both slices as witnesses and validate their sum equals limb1
-        field_t limb_1_lo = field_t::from_witness(ctx, limb_1_lo_v);
-        field_t limb_1_hi = field_t::from_witness(ctx, limb_1_hi_v);
-
-        // We need to propagate the origin tag to the chunks of limb1
-        limb_1_lo.set_origin_tag(limb1.get_origin_tag());
-        limb_1_hi.set_origin_tag(limb1.get_origin_tag());
-        limb1.assert_equal((limb_1_hi * limb_1_hi_multiplicand) + limb_1_lo);
-
-        // Step 4: apply range constraints to validate both slices represent the expected contributions to *this.lo and
-        // *this,hi
-        limb_1_lo.create_range_constraint(lo_bits_in_limb_1);
-        limb_1_hi.create_range_constraint(hi_bits_in_limb_1);
-
-        // construct *this.lo out of:
-        // a. `limb0` (the first NUM_LIMB_BITS bits of scalar)
-        // b. `limb_1_lo` (the first LO_BITS - NUM_LIMB_BITS) of limb1
-        lo = limb0 + (limb_1_lo * BigScalarField::shift_1);
-
-        const uint256_t limb_2_shift = uint256_t(1) << (BigScalarField::NUM_LIMB_BITS - lo_bits_in_limb_1);
-        const uint256_t limb_3_shift =
-            uint256_t(1) << ((BigScalarField::NUM_LIMB_BITS - lo_bits_in_limb_1) + BigScalarField::NUM_LIMB_BITS);
-
-        // construct *this.hi out of limb2, limb3 and the remaining term from limb1 not contributing to `lo`
-        hi = limb_1_hi.add_two(limb2 * limb_2_shift, limb3 * limb_3_shift);
+        lo.set_origin_tag(scalar.get_origin_tag());
+        hi.set_origin_tag(scalar.get_origin_tag());
+        return;
+    }
+
+    // Step 1: Ensure the bigfield scalar fits into LO_BITS + HI_BITS by reducing if necessary.
+    // Note: we can tolerate the scalar being > ScalarField::modulus, because performing a scalar mul implicitly
+    // performs a modular reduction.
+    if (scalar.get_maximum_value() >= (uint512_t(1) << (LO_BITS + HI_BITS))) {
+        scalar.self_reduce();
```

Comment:
The diff here is not nearly as large as github makes it look. I haven't changed the logic in this method in any fundamental way - just some renaming and rewriting of the comments for improved clarity.